src/stats.c

```diff
@@ -0,0 +1,364 @@
+ * (divides by [size - 1] since the given values are considered to be a sample
+ * of the population)
+ * @param array array containing the values of the sample
+ * @param size  number of samples
+ * @return absolute average deviation of sample
+ */
+real_number absAverageDeviation(real_number* array, vec_size size) {
+  real_number avg = mean(array, size);
+  real_number result = 0.0;
+  for (vec_size i = 0; i < size; ++i) {
+    real_number elem = array[i];
+    real_number diffElem = elem - avg;
+    result += fabs(diffElem);
+  }
+
+  return result / (size - 1);
+}
+
+/**
+ * @brief Computes the absolute average deviation of a sample
+ * (divides by [size - 1] since the given values are considered to be a sample
+ * of the population)
+ * @param array array containing the values of the sample
+ * @param size  number of samples
+ * @return absolute average deviation of sample
+ */
+real_number covariance(real_number* x, real_number* y, vec_size size) {
+  real_number avgX = mean(x, size);
+  real_number avgY = mean(y, size);
+
+  real_number sum = 0.0;
+  for (vec_size i = 0; i < size; ++i) {
+    sum += (x[i] - avgX) * (y[i] - avgY);
+  }
+
+  return sum / (size - 1);
+}
+
+/**
+ * @brief Makes a statistical analysis of an array of numbers
+ * @param array     array containing the data
+ * @param size      size of the array
+ * @param sample    whether to consider the data as a sample or as the
+ * population when computing the metrics (0 = Population, 1 = Sample)
+ * @return Returns the error code of the function (0 = Success, 1 = Size of data
+ * is less than 2, 2 = Variance is null, can't compute skewness and kurtosis,
+ * but other metrics are computed)
+ */
+vec_size analyzeData(real_number* array, vec_size size, DataSummary* moments) {
+
+  if (size < 2)
+    return 1;
+
+  real_number avg = mean(array, size);
+  real_number max = array[0], min = array[0], absAverageDeviation = 0.0,
+              var = 0.0, skew = 0.0, kurtosis = 0.0;
+  real_number adujstingFactor = 0.0;
+  for (vec_size i = 0; i < size; ++i) {
+    real_number elem = array[i];
+
+    if (max < elem)
+      max = elem;
+    if (min > elem)
+      min = elem;
+
+    real_number diffElem = elem - avg;
+    real_number diffElemSquared = diffElem * diffElem;
+    absAverageDeviation += fabs(diffElem);
+    adujstingFactor += diffElem;
+    var += diffElemSquared;
+    skew += diffElemSquared * diffElem;
+    kurtosis += diffElemSquared * diffElemSquared;
+  }
+
+  moments->min = min;
+  moments->max = max;
+  moments->mean = avg;
+
+  real_number dividingFactor = 1.0 / (size - 1);
+
+  moments->absAverageDeviation = absAverageDeviation * dividingFactor;
+  moments->variance =
+      (var - adujstingFactor * adujstingFactor / size) * dividingFactor;
+  moments->standardDeviation = sqrt(moments->variance);
+
+  if (fabs(moments->variance) < EPSILON) {
+    return 2; // Cannot compute skewness and kurtosis if variance is 0
```

Comment:
Maybe it would be good to define error codes in the header file and to use them here

Reply:
Yess, good call